test/functional/createnewblock.py

```diff
@@ -0,0 +1,322 @@
+
+    def test_recent_transactions(self):
+        # Populate the mempool with some transactions, and compare gbt results
+        # across the nodes.  Repeat.
+        self.add_empty_block(self.nodes[0])  # Bypass gbt cache.
+        self.sync_all()
+        recent_tx_block_count = [0, 0, 0]
+
+        max_iterations = 20
+        for i in range(max_iterations):
+            self.populate_mempool(random.randint(5, 15))  # 5-15 transactions
+
+            # Freeze current time, so that our recent tx window calculations
+            # will be consistent.
+            cur_time = int(time.time())
+            [node.setmocktime(cur_time) for node in self.nodes]
+
+            # Check that total block reward is within stale_rate of node3
+            max_income = self.nodes[3].getblocktemplate({
+                "rules": ["segwit"]
+                })['coinbasevalue']
+
+            # Check that calling gbt on the given node produces a block
+            # within the expected tolerance of max_income.
+            # Return true if the returned block includes recent transactions;
+            # false otherwise
+            def check_gbt_results(node, max_income, cnb_args, cur_time):
+                template = node.getblocktemplate({"rules": ["segwit"]})
+                mempool = node.getrawmempool(verbose=True)
+                stale_rate = cnb_args['stalerate']
+                # Convert window from milliseconds to seconds
+                window = (cnb_args['window'] + 999) // 1000
+
+                # We should always be within stale_rate of best possible block
+                assert template['coinbasevalue'] >= stale_rate * max_income
+
+                # Check that if we're including any recent transactions, then
+                # we achieve the max_income
+                includes_recent_transactions = False
+                for x in template['transactions']:
+                    if mempool[x['txid']]['time'] > cur_time - window:
+                        includes_recent_transactions = True
+                        break
+
+                if includes_recent_transactions:
+                    assert template['coinbasevalue'] == max_income
+
+                return includes_recent_transactions
+
+            for j in range(3):
+                if check_gbt_results(self.nodes[j], max_income,
+                                     self.cnb_args[j], cur_time):
+                    recent_tx_block_count[j] += 1
+
+            # Drain the mempool if it gets big
+            if self.nodes[0].getmempoolinfo()['bytes'] > 400000:
+                self.nodes[0].generate(1)
+            else:
+                [node.setmocktime(0) for node in self.nodes]
+                self.add_empty_block(self.nodes[0])
+            self.sync_all()
+
+        if sum(recent_tx_block_count) == 3 * max_iterations:
+            self.log.warn(
+                "Warning: every block contained recent transactions!"
```

Comment:
In commit "[qa] Add functional test for CreateNewBlock"

Does it make sense to warn if no block contains recent transactions?

Reply:
Done in 497ae4844922caf4e2244b847a5a63fd2d1a999a